src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -202,6 +202,9 @@ private void SetGlobalErrorIfTrue(bool arg)
                     new LocalizableResourceString(messageResourceName, CodeAnalysisResources.ResourceManager, typeof(CodeAnalysisResources)));
             }
 
+            addCircularStructDiagnostics(compilation.GlobalNamespace);
```

Comment:
>compilation.GlobalNamespace [](http://example.com/codeflow?start=41&length=27)

It looks like we are traversing things from added modules and referenced assemblies. See ```methodCompiler.CompileNamespace``` call above.

Reply:
Yeah `compilation.GlobalNamespace` is **alot** of unnecessary work. I'm switching to `compilation.SourceModule.GlobalNamespace`. Thanks!

Reply:
> Yeah compilation.GlobalNamespace is alot of unnecessary work. 

It not only a lot of unnecessary work, but also the work that is likely to have negative impact on referenced compilations (work done out of "order").